retrieval/discovery/azure.go

```diff
@@ -0,0 +1,185 @@
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ad.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// azureClient represents multiple Azure Resource Manager providers.
+type azureClient struct {
+	nic network.InterfacesClient
+	vm  compute.VirtualMachinesClient
+}
+
+// createAzureClient is a helper function for creating an Azure compute client to ARM.
+func createAzureClient(cfg config.AzureSDConfig) (azureClient, error) {
+	var c azureClient
+	oauthConfig, err := azure.PublicCloud.OAuthConfigForTenant(cfg.TenantID)
+	if err != nil {
+		return azureClient{}, err
+	}
+	spt, err := azure.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, cfg.ClientSecret, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return azureClient{}, err
+	}
+
+	c.vm = compute.NewVirtualMachinesClient(cfg.SubscriptionID)
+	c.vm.Authorizer = spt
+
+	c.nic = network.NewInterfacesClient(cfg.SubscriptionID)
+	c.nic.Authorizer = spt
+
+	return c, nil
+}
+
+func (ad *AzureDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{}
+	client, err := createAzureClient(*ad.cfg)
+	if err != nil {
+		return tg, fmt.Errorf("could not create Azure client: %s", err)
+	}
+
+	var machines []compute.VirtualMachine
+	result, err := client.vm.ListAll()
+	if err != nil {
+		return tg, fmt.Errorf("could not list virtual machines: %s", err)
+	}
+	machines = append(machines, *result.Value...)
+	// if we still have more results, keep going
+	for result.NextLink != nil {
+		result, err = client.vm.ListAllNextResults(result)
+		if err != nil {
+			return tg, fmt.Errorf("could not list virtual machines: %s", err)
+		}
+		machines = append(machines, *result.Value...)
+	}
+	log.Infof("Found %d virtual machines during Azure discovery.", len(machines))
+
+	// We have the slice of machines. Now turn them into targets.
+	// Doing them in go routines because the network interface calls are slow.
+	var wg sync.WaitGroup
+	targets := make([]model.LabelSet, len(machines))
+	for i, vm := range machines {
+		wg.Add(1)
+		go func(i int, vm compute.VirtualMachine) {
+			defer wg.Done()
+			resource_group := strings.ToLower(strings.Split(*vm.ID, "/")[4])
+			labels := model.LabelSet{
+				azureLabelMachineID:       model.LabelValue(*vm.ID),
+				azureLabelMachineName:     model.LabelValue(*vm.Name),
+				azureLabelMachineLocation: model.LabelValue(*vm.Location),
+				azureLabelMachineRG:       model.LabelValue(resource_group),
+			}
+
+			// Get the IP address information via seperate call to the network provider.
+			for _, nic := range *vm.Properties.NetworkProfile.NetworkInterfaces {
+				group := strings.Split(*nic.ID, "/")[4]
+				name := strings.Split(*nic.ID, "/")[8]
+				networkInterface, err := client.nic.Get(group, name, "")
+				if err != nil {
+					log.Errorf("Unable to get network interface %s: %s", name, err)
```

Comment:
All other comments should be addressed except for this one. I'm not sure how to best error out from inside a routine. I'm still new to the Go language.

Would a panic here be acceptable?

Otherwise, I was thinking about making a mutex protected bool like 'errorFound' which defaults to False and gets set to True in any routine with a failure. The specifics would be logged to Errorf and I'd return an error from the refresh() function.